### ISSUE TITLE
Are transpilers adding + core.js invalid?

### ISSUE THREAD

**Issue by aikeru** [compat-table/compat-table#581]
Recently TypeScript's compatibility score was updated and includes + core.js. Perhaps I misunderstand, but this doesn't seem accurate.

Since core.js is a library, in theory, I could just add it to any environment as I would any shim/shiv and get a boost or a bonus for compatibility from that library.

So, if TypeScript + core.js is valid, why isn't this applied for all other environments ... ie: Chrome + core.js?

If core.js was a dependency of TypeScript or otherwise directly integrated (as opposed to a 3rd party library) then TypeScript would probably get the compatibility score automatically but as I understand it, this isn't the case.

IMO, TypeScript 1.5-alpha should be represented without core.js. If core.js would really be added, it should be the same as using flags or experimental stuff in Chrome - put the semi-transparent secondary score and note that this is "with core.js" (again, why don't all have this?).

I think this would be a more consistent approach. Please educate me if I've missed something, thanks.

**Comment by ljharb** [comment]
Agreed - the same applies to `es6-shim` (where it might be nice to see "Chrome + es6-shim" and "Firefox + es6-shim" etc). Also, should "babel + core-js" even mention core-js, since it comes with babel? Should "core-js" get its own column by itself?

**Comment by teppeis** [comment]
Similar issue: https://github.com/Microsoft/TypeScript/issues/3956

**Comment by RyanCavanaugh** [comment]
The comparison to "Chrome + core.js" is irrelevant. Chrome is a _runtime_ and needs to support these things natively -- the web is not going to move forward if every browser just ships a slow Map polyfill and calls it a day.

The core principle is that all of the transpilers should be scored on equal footing in terms of having a polyfill library available at runtime. Transpilers that _require_ you to have another script library present at runtime shouldn't earn bonus points.

Your suggestion that it matter whether or not a transpiler takes a direct dependency on core.js creates perverse incentives. If tomorrow Tracuer said "Hey guys our download is now 54 kb bigger and we've done literally nothing else, so we're going to raise our score by 20 points", that's just making the download bigger and not actually improving anyone's life (presumably anyone capable of using a JS compiler is also capable of downloading a polyfill library without help from someone else). Conversely, wouldn't Babel be _better_ if its core.js dependency were totally optional -- at which point we would _lower_ its score? That doesn't make sense either.

I question the relevancy of whether or not TypeScript includes core.js as part of its dependency chain. If TypeScript includes core.js in a folder somewhere in its NPM download, does that actually _change_ TypeScript in any meaningful way? Is the developer experience actually affected by downloading 1 extra library in addition to the 20 they're probably already using anyway? Does it matter to the person downloading the webpage whether the core.js script they have to load came from the transpiler's distribution or from a bower package?

**Comment by aikeru** [comment]
@RyanCavanaugh 
Thanks for your reply. We'll have to agree to disagree on whether including core.js + a browser is relevant or not. Saying TypeScript + core.js may as well say "TypeScript + any ES6 compliant browser + type definitions for the same".

As it is, core.js is - well - "core" to much of what babel.js provides out of the box, batteries included. 
What you're saying is like claiming that TypeScript has all of the features of jQuery - you just need to include the jQuery library AND the type definitions and you're good to go! That would be silly, right?

I think you missed some of the points I've made here. The first of which, is that the TypeScript folks could simply build a better transpiler - for example, only shim features that are actually used by the code being transpiled. Or, make core.js or its shimmed features it an opt-in feature - of TypeScript!

As it is, TypeScript's "core" mission IS NOT to be an ES6/7 transpiler first, rather it is to be a typed superset of JavaScript. So, naturally, something like Babel, which is designed first and foremost and purpose-built to give as much ES-next features as possible, will have a higher score and include overhead that TypeScript wont.

Whether or not core.js is included is completely relevant. Go to the TypeScript playground and try to use Array.from, for example. Yes, the TypeScript compiler will ignore it and your code will still execute albeit with a red squiggly line IF the underlying browser runtime supports it. If you do the same thing with Babel, it will work regardless of browser runtime AND it wont break your CI build because you didn't add type definitions.

Further, implying this automatically requires including a 54kb library would show a lack of imagination on how to solve this problem.

Let's make this really, really simple.
What percentage of ES6 (or 7, etc.) works with Babel out of the box, in the default configuration?
What percentage of ES6 (or 7, etc.) works with TypeScript out of the box, in the default configuration?

Babel and TypeScript both transpile classes in different ways. There will be differences in overhead in the approaches both transpilers take. You may think it's better if babel didn't include core.js, but that's the great thing about have a community of choices. Babel folks get all that functionality up front and TypeScript doesn't, though TypeScript has lower overhead. Hurrah!

Do you not see the inconsistency in your argument that transpilers should be measured on equal footing without external libraries and yet you argue in favor of including "+ core.js" for TypeScript?
That could only be close to being "fair" on the day Microsoft's TypeScript website not only links to core.js AND the ES6 type definitions required, but includes step-by-step instructions on how to use it with TypeScript.

**Comment by aikeru** [comment]
I feel that this new title and focus is more helpful.

**Comment by ljharb** [comment]
@aikeru the title, yes, but I don't think this should be closed just yet. Reopening.

**Comment by siemato** [comment]
I think that both Babel+corejs and TypeScript+corejs are misleading. Some features in ES6/7/next are getting transpiled down to ES5, some features are supported (but don't get transpiled to lower ES versions), and in addition to that, it's getting more and more important, to see, which features are NOT supported, instead of 'can be achieved' by adding shims/core-js/... - Is it possibly a better choice to split it up Babel, TypeScript and core-js into seperate rows?

**Comment by chicoxyzzy** [comment]
agree 💯%
